### ISSUE TITLE
Avoid external platform headers (windows.h, X11/Xlib.h, ...)

### ISSUE THREAD

**Issue by pc-john** [KhronosGroup/Vulkan-Docs#2025]
As discussed originally at KhronosGroup/Vulkan-Headers#234, there are number of issues stemming from the includes of platform headers such as windows.h inside vulkan.h or one of its sub-headers.

The problems caused by the includes of external platform headers (windows.h, X11/Xlib.h, etc.) fall into two categories:
- **namespace pollution** - windows.h defines thousands and thousands of macros and symbols that clash with user's code. The macros include, for example, min, max, near, far and MemoryBarrier, making it difficult even to use stl library (e.g. std::max, or std::numeric_limits<uint64_t>::max(), .... because these clash with windows.h defined macros. Even vk_sdk_platform.h have to define NOMINMAX and undef inline and the header vulkan.hpp have to undef MemoryBarrier. Using variables like near and far is not possible in your projects without further #undefs. Other difficulties and name clashes are probably waiting just behind the doors. The same story repeats for X11/Xlib.h. Defining types like Window or Display makes your project fail to compile because these are defined by X11. Overall, I consider the inclusion of external platform headers not good idea.
- **extended compilation time** - particularly windows.h includes enormous number of subheaders, resulting in 350ms overhead on my i7-10875H. It might not seem too much, but for large projects it might be considerable overhead. 1000 files already results in 11 minutes of CPU time.

The ideal solution for a naive user would be:
#include <vulkan/vulkan.h>
and that's it. Nothing more required.

To get rid of external platform headers, we can use:
#include <vulkan/vk_platform.h>
**#include <vulkan/vulkan_core.h>**
#ifdef VK_USE_PLATFORM_WIN32_KHR
typedef struct HINSTANCE__* HINSTANCE;
typedef struct HWND__* HWND;
typedef struct HMONITOR__* HMONITOR;
typedef void* HANDLE;
typedef /*_Null_terminated_*/ const wchar_t *LPCWSTR;
typedef unsigned long DWORD;
typedef struct _SECURITY_ATTRIBUTES SECURITY_ATTRIBUTES;
**#include <vulkan/vulkan_win32.h>**
#endif
#ifdef VK_USE_PLATFORM_XLIB_KHR
typedef struct _XDisplay Display;
typedef unsigned long Window;
typedef unsigned long VisualID;
**#include <vulkan/vulkan_xlib.h>**
#endif
#ifdef VK_USE_PLATFORM_XCB_KHR
struct xcb_connection_t;
typedef uint32_t xcb_window_t;
typedef uint32_t xcb_visualid_t;
**#include <vulkan/vulkan_xcb.h>**
#endif
#ifdef VK_USE_PLATFORM_WAYLAND_KHR
**#include <vulkan/vulkan_wayland.h>**
#endif
#ifdef VK_USE_PLATFORM_XLIB_XRANDR_EXT
typedef unsigned long RROutput;
**#include <vulkan/vulkan_xlib_xrandr.h>**
#endif
#ifdef VK_USE_PLATFORM_VI_NN
**#include <vulkan/vulkan_vi.h>**
#endif
 
As you can see, it is quite a piece of code to get rid of the external platform headers while it would be even longer as there are still some platforms missing.

Are users expected to do all this? Can we help him already by better design of Vulkan headers? Can the current situation be improved?

Some thoughts, ideas and proposals:

- **no external platform headers** - we might not include external platform headers in vulkan.h and define all necessary symbols in vulkan.h. Unfortunately, this would break the current Vulkan spec as these have to be included. Anyway, can this be an option for Vulkan 2.0 ? Or can we avoid breaking the spec by introducing alternative vulkan.h header, such as vulkan_wsi.h or vulkan_platforms.h that would define the required symbols instead of including the external platform headers?
- **vulkan_wayland.h approach** - we might define all the required types "inline", like vulkan_wayland.h is doing:
typedef struct VkWaylandSurfaceCreateInfoKHR {
    VkStructureType                   sType;
    const void*                       pNext;
    VkWaylandSurfaceCreateFlagsKHR    flags;
    **struct wl_display***                display;
    **struct wl_surface***                surface;
} VkWaylandSurfaceCreateInfoKHR;
This way, we might make includes of external platform headers optional. Users might omit them and Vulkan headers would still compile fine.

Does anything of this make sense? Do you see some arguments against all this? Or, do you see better ways around?

**Comment by krOoze** [comment]
There was  previously mechanism added to achieve this. You can include `vulkan_win32.h`, in which case you take on the responsibility (and resulting problems) of how `windows.h` stuff is incorporated.

**Comment by mackron** [comment]
Yep, inlining the Win32 types vulkan.h as proposed in the original post will just break builds that use windows.h in the same translation unit. I personally build everything as one big translation unit and require windows.h anyway for things like window creation, and I'm certainly not the only one who'd be calling `CreateWindowEx()` in the same translation unit that's also referencing Vulkan.

The only valid solution is to have a mechanism that allows people to disable inclusion of system headers inside vulkan.h, maybe via a `#define` or something, and then require those who enable that option to include the necessary platform-specific definitions before their vulkan.h include.

**Comment by oddhack** [comment]
IIRC you were getting the platform header as a result of using the Vulkan-Hpp C++ headers, which don't give the granularity of control to include or not include the WSI headers? If that's the case I still say you should ask Vulkan-Hpp about providing that control, and perhaps also propose a PR that would do what you're suggesting at the C header level and try to get other folks using that approach as a proof of concept.

**Comment by pc-john** [comment]
> Yep, inlining the Win32 types in vulkan.h as proposed in the original post will just break builds that use windows.h in the same translation unit.

I am not C++ guru, so I wrote few tests yesterday and this morning. I see no problems at all. If we take Win32 platform, for example, we shall be able to change:

<pre>typedef struct VkWin32SurfaceCreateInfoKHR {
    VkStructureType                 sType;
    const void*                     pNext;
    VkWin32SurfaceCreateFlagsKHR    flags;
    <b>HINSTANCE                       hinstance;
    HWND                            hwnd;</b>
} VkWin32SurfaceCreateInfoKHR;</pre>

to

<pre>typedef struct VkWin32SurfaceCreateInfoKHR {
    VkStructureType                 sType;
    const void*                     pNext;
    VkWin32SurfaceCreateFlagsKHR    flags;
    <b>struct HINSTANCE__*             hinstance;
    struct HWND__*                  hwnd;</b>
} VkWin32SurfaceCreateInfoKHR;</pre>

Another "advanced" structure would be converted like this:

<pre>typedef struct VkExportMemoryWin32HandleInfoKHR {
    VkStructureType               sType;
    const void*                   pNext;
    <b>const SECURITY_ATTRIBUTES*    pAttributes;
    DWORD                         dwAccess;
    LPCWSTR                       name;</b>
} VkExportMemoryWin32HandleInfoKHR;</pre>

to

<pre>typedef struct VkExportMemoryWin32HandleInfoKHR {
    VkStructureType               sType;
    const void*                   pNext;
    <b>const struct _SECURITY_ATTRIBUTES*    pAttributes;
    unsigned long                         dwAccess;
    /*_Null_terminated_*/ const wchar_t*  name;</b>
} VkExportMemoryWin32HandleInfoKHR;</pre>

Such code is indifferent to <code>#include <windows.h></code>, at least based on my tests. If <code>windows.h</code> is included before this code, it does not disturb the declaration above. If it is not, it compiles the same. If it is included later, after this structure, it still compiles fine.

> I personally build everything as one big translation unit and require windows.h anyway for things like window creation, and I'm certainly not the only one who'd be calling `CreateWindowEx()` in the same translation unit that's also referencing Vulkan.

I use different approach. I put all my platform dependent code in one or only few source files. These files include <code>windows.h</code>, <code>X11/Xlib.h</code>, etc. These files pay the full price of "damage" caused by includes of these headers. The rest of my project is usually platform independent code, enjoying all the usual benefits of programming freedom :-).

In conclusion, I understand the risk of changing whatever in the Vulkan headers as it affects the whole world. Anyway, we have the precedent of Wayland platform (#1905) that already uses the approach suggested above. Thoughts?

**Comment by mackron** [comment]
It's working in your case because the compilers you're using happen to be using a consistent way of defining each type. The problem is that not all compilers ship with headers that do the same thing. Digital Mars, for example, complains with this:
```
typedef struct HINSTANCE__* HINSTANCE;
                                     ^
test_deviceio/ma_test_deviceio.c(48) : Error: 'HINSTANCE' previously declared as something else
It was declared as: void *
It is now declared: struct HINSTANCE__*
```
You can see from the error that DM is just using `void *`. In my opinion, by redeclaring types from system headers you're really opening a can of worms when it comes to compiler compatibility. Using windows.h directly is just safer.

Having said all this, I fully support the overall idea of what you're saying, it's just your specific proposal in the original post I don't like is all. I absolutely think there should be some kind of mechanism to disable automatic inclusion of platform-specific headers and allow people to declare the necessary types themselves.

**Comment by krOoze** [comment]
As I said, there already is mechanism to disable inclusion of platform headers:

```
#include <vulkan/vulkan_core.h>
#include <vulkan/vulkan_win32.h>
```

**Comment by pc-john** [comment]
> Digital Mars, for example, complains [...]

Thanks! This is real argument. It shows that the path might not be so straghtforward. Surely, we would also need to test the code on Clang, Intel, MinGW, Cygwin, maybe also Borland compilers (?), and others. I expect that some number of compilers will go Microsoft way. The "void*" of Digital Mars is the other "natural" way. Could we expect many other ways? Theoretically yes, but I have some doubts that there are many kinds of windows headers that are actively maintained.

In conclusion, I understand your worries about opening can of worms. The defensive approach should probably be used. Maybe, to develop the solution for MSVC only with the possibility to expand to more compilers in the future, while probably keeping all the work behind some define (?) until it is considered well tested.

So, the solution would be:
<pre>
#ifndef VK_SKIP_PLATFORM_HEADERS

typedef struct VkWin32SurfaceCreateInfoKHR {
    VkStructureType                 sType;
    const void*                     pNext;
    VkWin32SurfaceCreateFlagsKHR    flags;
    <b>HINSTANCE                       hinstance;
    HWND                            hwnd;</b>
} VkWin32SurfaceCreateInfoKHR;

[... all other structs and stuff ...]

#elif defined(_MSC_VER) || defined(__otherMSVCcompatibleCompiler__)

typedef struct VkWin32SurfaceCreateInfoKHR {
    VkStructureType                 sType;
    const void*                     pNext;
    VkWin32SurfaceCreateFlagsKHR    flags;
    <b>struct HINSTANCE__*             hinstance;
    struct HWND__*                  hwnd;</b>
} VkWin32SurfaceCreateInfoKHR;

[... all other structs and stuff ...]

#elif defined(__DMC__)

typedef struct VkWin32SurfaceCreateInfoKHR {
    VkStructureType                 sType;
    const void*                     pNext;
    VkWin32SurfaceCreateFlagsKHR    flags;
    <b>void*                           hinstance;
    void*                           hwnd;</b>
} VkWin32SurfaceCreateInfoKHR;

[... all other structs and stuff ...]

#endif
</pre>

There might be other approaches, but this one seems to me to provide the least namespace pollution.

I am guessing that the same issue on Xlib would be much simpler. I do not believe that different compilers provide different Xlib headers, so it shall be much simpler there.

**Comment by mackron** [comment]
@pc-john That might work, but then you're polluting the headers which not only makes it extremely messy, but unnecessarily difficult to maintain. What happens when a compiler comes along which does something different again, like `struct __HWND*` or `struct _HWND_*` or `struct __WINDOW_HANDLE__*`, etc., etc.? What happens if Digital Mars comes along and say "Hey, our HWND define is different to GCC, Clang and MSVC - we better update our headers so they're compatible!" and then they make a change which breaks the Vulkan headers? It's just a mess all round in my opinion.

@krOoze So the idea is that you would do something like this?
```c
#include "my_custom_windows.h"
#include <vulkan/vulkan_core.h>
#include <vulkan/vulkan_win32.h>
```
What happens if I want to add support for other platforms? I need to add the headers for other platforms?
```
#include <vulkan/vulkan_xlib.h>
```
This is an awful solution. It doesn't scale as new platforms are added. It's inconsistent with documentation and examples which just causes unnecessary complexity, fragmentation and confusion. It's not drop-in compatible with projects like mine which amalgamate the Vulkan headers into one big header for simplified deployments. It's just generally unhygienic. There should be one standard way to include everything you need to use Vulkan - vulkan.h. Anything else just causes unnecessary fragmentation.

In my opinion, the base way to do it should be to include vulkan.h like normal:
```
#include <vulkan/vulkan.h>
```
Then I decide you know what, I don't like the windows.h namespace pollution and increased compile times, I'm going to do a Win32-specific optimization:
```
#define VK_NO_WINDOWS_H
#include "my_custom_windows.h"
#include <vulkan/vulkan.h>
```
This way everything is still using the standard vulkan.h which is consistent with all other C/C++ Vulkan code all over the internet. I'm just augmenting it with a Win32-specific optimization. The non-Win32 builds are totally unaffected by the change. Don't even need to think about them. And as a bonus, If I decide that I want to replace the Khronos-generated vulkan.h with a drop-in replacement I can do that.

**Comment by krOoze** [comment]
> There should be one standard way to include everything you need to use Vulkan - vulkan.h. 

And so there it is there. But if you have special needs, you do have special ways.

> It's just generally unhygienic.

IMO macro-configurable monolithic headers **are** unhygienic IMO. Such as is the `windows.h`, which is the whole source of the problems here because of it. If we could simply include only what we want from `windows.h` without messy macros that would be great.

Consider this:

```
\\ a.h
#define VK_NO_WINDOWS_H
#include <vulkan/vulkan.h>

\\ b.h
#define VK_YES_WINDOWS_H
#include <vulkan/vulkan.h>
void f( WindowsType t );

\\ c.c
#include "a.h"
#include "b.h"
```

This is IMO unhygienic.

**Comment by mackron** [comment]
I don't think my `VK_NO_WINDOWS_H` suggestion is really much different to the `VK_USE_PLATFORM_*` thing. Obviously if you're including vulkan.h with different options it's going to be erroneous. In my mind you would define `VK_NO_WINDOWS_H` alongside your `VK_USE_PLATFORM_*` define (be it in-source or in your build script) and then the only difference would be your custom Win32 defines ("my_custom_windows.h" in my examples). I don't think it's *that* messy (not ideal though).

I don't know. It's all personal preference I guess. I just sit firmly in the "just have one file" camp (I was also whinging about the vk_video headers that were added recently!).

But anyway, I've said my piece and I'll leave it there.

**Comment by cubanismo** [comment]
Agree with @oddhack here. We fixed the C header files to give apps the ability to segment their system header includes themselves and cordon off the platform-specific Vulkan stuff if they so desire. Wayland was further cleaned up because (1) its few platform-specific types are trivial to forward-declare and all referenced only as pointers, and (2) we ran into a mess where its header files introduce piles of warnings with newer compilers, but otherwise I'm pretty happy with the current balance. If vulkan.hpp hasn't been similarly updated to expose equivalent granularity, that seems like something someone working on that project should tackle.

> Are users expected to do all this?

Frankly, no. The way we expected developers to use these platform-specific Vulkan headers was to write your platform-specific Vulkan glue code in a small translation unit that includes the native headers, perhaps with your favorite windows.h behavior-affecting macros like lean-and-mean, but if you want to go through the trouble and type out definitions of the types required as above, that's fine too. Vulkan has never strove to be an API that strives to minimize typing AFAICT. Users are always free to write their own my_no_platform_headers_vulkan.h and proselytize its use as they see fit, but I still think the current Vulkan headers strike the right balance between maintainability, minimalism, and a backwards-compatible vulkan.h.

**Comment by pc-john** [comment]
> The way we expected developers to use these platform-specific Vulkan headers was to write your platform-specific Vulkan glue code in a small translation unit that includes the native headers

I agree. If I understand you correctly: to concentrate the platform specific code in just one or few translation units. This way, the rest of project does not need to know anything about windows.h, X11/Xlib.h,.... Unfortunately, a project might use dynamic Vulkan function pointer loading. The table with function pointers might be placed in a header that is included by almost whole project. Unfortunately, the table might also contain the platform specific function pointers. And that is where we did not want to be: Because of the platform specific pointers, the whole project will need to include platform specific vulkan headers and system headers (windows.h,...).

I see two workarounds for this: Either to provide some placeholders (such as void*) into the table and typecast it to proper types in platform specific translation units. Although not nice, it would work.

For completeness, there might be also another approach. For instance Qt avoids platform specific types in most important classes by using [WId type](https://doc.qt.io/qt-6/qwindow.html#winId), which is essentially size_t or similar 8 byte integer type so it can hold pointer. So, instead of platform specific HWND, they use their platform neutral WId, which serves for HWND on Windows, for Window type on Xlib, etc. I am not saying that we must go this way, but anyway, what would you say on similar approach in Vulkan headers? I mean to replace all platform specific types by, for instance, VkPlatformHandle which would be void* or size_t, or so. Surely, all this shall be guarded by some define or something to not break backward compatibility. It would be users responsibility to cast it to the proper type while Vulkan headers would be completely platform types agnostic. With such approach, we might forget on any vulkan_[platform].h and all VK_USE_PLATFORM_[platform]_KHR macros, having finally just <code>#include <vulkan/vulkan.h></code>. I understand how radical the change is, but I like how it massively simplifies the things. At least I hope so. Even if it would be considered too radical for Vulkan 1.x, it might be considered as idea for Vulkan 2.0. Even in the case it will not be used in the end, I think it is good to put all ideas and arguments in this thread for consideration here and for future reference of other people having the same ideas or requests.

**Comment by oddhack** [comment]
@pc-john after some internal discussion, the feeling of the Vulkan WG is that we've explained why it is the way it is, and offered a couple of options that may help - ask Vulkan-Hpp to accommodate you, and/or propose a PR here so other people can evaluate what you're asking for in the C headers. But we're not going to ourselves put in a significant header change like this - it would take a lot of testing to ensure nobody will be negatively affected by it.

If you do want to propose a PR, you can link it off this issue, but otherwise, please close this.

**Comment by pc-john** [comment]
@oddhack thanks for your consideration in WG! I appreciate that you take the community seriously! From my side, I think, all was said and thanks for your time. I think, I understand your position and the reasons. 

Vulkan-Hpp biggest problem is long compilation time (~4 seconds, guessing). So, we are currently running a research on this particular topic with the possibility of replacing Vulkan-Hpp with a new C++ binding generator. The new generator would surely include Vulkan C headers. That is the reason of my special interest in quality of C headers. Thanks for all your work here! Appreciate it.

Thanks also for offering the PR. I will close the issue if there will be no more comments from others in the next two days. If any PR will be made by me, I will link it with this issue. 

If anybody from internet would be interested in seeing something of this to materialize, please mark the post with "Thumb up". There is no reason to bother Vulkan WG with whatever if others will not profit of it.
